### DISCUSSION TITLE
HELICS in multi computer

### DISCUSSION THREAD

**Discussion by giuseppelazzara98** [GMLC-TDC/HELICS#2576] [General]
Hello everyone! I am running HELICS on different computers and I am having problems registering federates. The configuration is as follows:

- 1 Broker and 1 federate on machine A

- 1 federate on a machine B

- 2 federates on machine C

All federates ones connect to the broker on machine A, but those found on machine B and C report this log.![image](https://github.com/GMLC-TDC/HELICS/assets/100286939/6a4a514c-e58d-4e6c-a481-6b31844d4bf5)

Instead the broker this log ![image](https://github.com/GMLC-TDC/HELICS/assets/100286939/b4edc7e3-a2e2-4810-b938-b3fb664ec539)

I can't understand what the problem is, it doesn't seem to be a network problem because the broker receives requests from the various federates but then can't register them

**Comment by phlptp** [comment]
what is the configuration you gave to the broker and to the federates?

**Reply by giuseppelazzara98** [reply]
This is the configuration for the machine that is running the broker and one federate:

```
std: :string initstring = "-f 2 --loglevel=debug --external";

std: string fedinitstring = "-brokeratcp://7.182.101.251 --federates=1 --loglevel=debug";

```

This is the configuration given to the other machines that are running the federates 

```
std: :string fedinitstring = "--broker=tcp://7.182.101.251 --federates=1 --log_level=debug ";
```
Very basic configuration.

**Reply by phlptp** [reply]
For the broker this should be std: :string initstring = "-f 4 --loglevel=debug --external";   The --federates=1 on the federates isn't going to do much, though I don't think it will impact anything.   But the -f 4 on the broker lets the broker know how many federates to expect.  If that number doesn't match the number of federates you expect the co-simulation could start before intended and not let the additional federates join.  

The only other similar thing I have seen is outbound firewall rules on the broker machine that prevent some of the communications from going through.    If that is the case, then using the zmq-ss core may resolve that as the port usage is a different pattern.

**Reply by giuseppelazzara98** [reply]
Yes, I forgot to specify it, I change that number(federates expected by the broker) at run-time, passing it as an argument to the executable.

Now I tried to use zmq_ss and I solved the problem.

Can this compromise performance instead of using the normal zmq?
Because I would like to use this in an environment where I need to take performance into account.

**Reply by phlptp** [reply]
The answer to that is somewhat nuanced and depends greatly on traffic patterns and other things.  The ZMQ_SS core was designed to more efficiently handle larger numbers of smaller federates on small number of machines.  In which case it dramatically outperforms the regular ZMQ core.    The regular zmq core uses a larger number of sockets in both directions which make the patterns more predictable, and prevents some forms of bottlenecking and is generally good in many cases which is why it is the default.  It also allows the priority messages to go through faster.    That being said in most use cases the performance is nearly identical and rarely is HELICS the limiting factor in performance.  If you actually run into performance issues let me know I would be very interested in it.

**Reply by giuseppelazzara98** [reply]
I am currently making benchmarks. I would also be happy to share them with you and I would love to have a feedback from you.
